src/compiler/transformers/esnext.ts

```diff
@@ -76,5 +100,409 @@ namespace ts {
+            return !!(node.transformFlags & TransformFlags.ContainsDefaultValueAssignments);
+        }
+
+        /**
+         * Adds statements to the body of a function-like node if it contains parameters with
+         * binding patterns or initializers.
+         *
+         * @param statements The statements for the new function body.
+         * @param node A function-like node.
+         */
+        function addDefaultValueAssignmentsIfNeeded(statements: Statement[], node: FunctionLikeDeclaration): void {
+            if (!shouldAddDefaultValueAssignments(node)) {
+                return;
+            }
+
+            for (const parameter of node.parameters) {
+                // A rest parameter cannot have a binding pattern or an initializer,
+                // so let's just ignore it.
+                if (parameter.dotDotDotToken) {
+                    continue;
+                }
+
+                if (isBindingPattern(parameter.name)) {
+                    addDefaultValueAssignmentForBindingPattern(statements, parameter);
+                }
+            }
+        }
+
+        /**
+         * Adds statements to the body of a function-like node for parameters with binding patterns
+         *
+         * @param statements The statements for the new function body.
+         * @param parameter The parameter for the function.
+         */
+        function addDefaultValueAssignmentForBindingPattern(statements: Statement[], parameter: ParameterDeclaration): void {
+            const temp = getGeneratedNameForNode(parameter);
+
+            // In cases where a binding pattern is simply '[]' or '{}',
+            // we usually don't want to emit a var declaration; however, in the presence
+            // of an initializer, we must emit that expression to preserve side effects.
+            if ((parameter.name as BindingPattern).elements.length > 0) {
+                statements.push(
+                    setEmitFlags(
+                        createVariableStatement(
+                            /*modifiers*/ undefined,
+                            createVariableDeclarationList(
+                                flattenParameterDestructuring(parameter, temp, visitor, /*transformRest*/ true)
+                            )
+                        ),
+                        EmitFlags.CustomPrologue
+                    )
+                );
+            }
+            else if (parameter.initializer) {
+                statements.push(
+                    setEmitFlags(
+                        createStatement(
+                            createAssignment(
+                                temp,
+                                visitNode(parameter.initializer, visitor, isExpression)
+                            )
+                        ),
+                        EmitFlags.CustomPrologue
+                    )
+                );
+            }
+        }

```

Comment:
Wow, that's an awful lot of code. I'm surprised it takes this much, but I'm not familiar enough with the new emitter to truly evaluate this. Make sure you have @rbuckton take a look.